### PR TITLE
AP_Baro:add driver for the SPA06

### PR DESCRIPTION
**Pull request by radiolinkW**
The SPA06 is a high resolution low noise barometer pin-to-pin compatible with the SPL06.
I have been flight testing this baro on RadiolinkPIX6 and driver it seems to work well.
Here is the datasheet: [SPA06-003_datasheet.pdf](https://github.com/radiolinkW/binary/blob/master/SPA06-003_datasheet.pdf)

### CODE REVIEW

**Comment by peterbarker** [comment]
Needs to compile

This looks *very* similar to the SPL06 driver.

This looks like we should be re-using the SPLP06 driver for the SPA, particularly if they can be distinguished by their WHOAMI registers or some other fingerprinting we can do.

We *seriously* can't afford to copy a driver sideways - we're fast running out of flash on boards, and baros make their way into a lot of boards.

Can you think about whether it would be possible to augment the SPL driver, please?

I'm happy to take this to a DevCall to see if other devs disagree, but I don't think they will.

**Comment by radiolinkW** [comment]
> Needs to compile
> 
> This looks _very_ similar to the SPL06 driver.
> 
> This looks like we should be re-using the SPLP06 driver for the SPA, particularly if they can be distinguished by their WHOAMI registers or some other fingerprinting we can do.
> 
> We _seriously_ can't afford to copy a driver sideways - we're fast running out of flash on boards, and baros make their way into a lot of boards.
> 
> Can you think about whether it would be possible to augment the SPL driver, please?
> 
> I'm happy to take this to a DevCall to see if other devs disagree, but I don't think they will.

I also think it should be made into a driver, but the class name should not be SPL06.
What do you think the class name should be?

**Comment by peterbarker** [comment]
> > Needs to compile
> > This looks _very_ similar to the SPL06 driver.
> > This looks like we should be re-using the SPLP06 driver for the SPA, particularly if they can be distinguished by their WHOAMI registers or some other fingerprinting we can do.
> > We _seriously_ can't afford to copy a driver sideways - we're fast running out of flash on boards, and baros make their way into a lot of boards.
> > Can you think about whether it would be possible to augment the SPL driver, please?
> > I'm happy to take this to a DevCall to see if other devs disagree, but I don't think they will.
> 
> I also think it should be made into a driver, but the class name should not be SPL06. What do you think the class name should be?SPX06？So the SPL06 driver needs to be rewritten.

Yes indeed, you could rename  `SPL06` to `SPx06` (lower-case x) and *only after creating a commit for that* adjust it to support the SPA06

**Comment by peterbarker** [comment]
To be clear, if you rename a file, to it in a completely separate commit to everything else, including adjusting `#include`s.  This ensures the history on the file is preserved.

**Comment by radiolinkW** [comment]
> To be clear, if you rename a file, to it in a completely separate commit to everything else, including adjusting `#include`s. This ensures the history on the file is preserved.

@peterbarker The SPA06 and SPL06 baro have been merged into an SPx06 driver and tested for narmal flight on the latest Copter4.5 branch.

File: libraries/AP_Baro/AP_Baro_SPx06.h [hunk 1]
```diff
@@ -2,23 +2,27 @@
...
 
-class AP_Baro_SPL06 : public AP_Baro_Backend
+class AP_Baro_SPx06 : public AP_Baro_Backend
 {
 public:
-    AP_Baro_SPL06(AP_Baro &baro, AP_HAL::OwnPtr<AP_HAL::Device> dev);
+	enum SPx06 {
+			BARO_SPL06 = 0,
+			BARO_SPA06 = 1,
+	};
```
Comment (peterbarker) [inline libraries/AP_Baro/AP_Baro_SPx06.h:24]:
```suggestion
	enum class Type {
	                UNKNOWN,
			SPL06,
			SPA06,
	};
```

... can do away with the is_spl06 and is_spa06 booleans after this change

File: libraries/AP_Baro/AP_Baro_SPx06.h [hunk 2]
```diff
@@ -45,7 +49,9 @@ class AP_Baro_SPL06 : public AP_Baro_Backend
 
     // Internal calibration registers
     int32_t _c00, _c10;
-    int16_t _c0, _c1, _c01, _c11, _c20, _c21, _c30;
+    int16_t _c0, _c1, _c01, _c11, _c20, _c21, _c30, _c31, _c40;
+
+    enum SPx06 _spx06_type;
```
Comment (peterbarker) [inline libraries/AP_Baro/AP_Baro_SPx06.h:54]:
```suggestion
    Type type;
```

File: libraries/AP_Baro/AP_Baro_SPx06.cpp [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+
+    _dev->set_speed(AP_HAL::Device::SPEED_HIGH);
+
+    uint8_t whoami;
+
+// Sometimes SPx06 has init problems, that's due to failure of reading using SPI for the first time. The SPx06 is a dual
+// protocol sensor(I2C and SPI), sometimes it takes one SPI operation to convert it to SPI mode after it starts up.
+    bool is_SPL06 = false;
+    bool is_SPA06 = false;
+
```
Comment (peterbarker) [inline libraries/AP_Baro/AP_Baro_SPx06.cpp:120]:
```suggestion
```

File: libraries/AP_Baro/AP_Baro_SPx06.cpp [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+    for (uint8_t i=0; i<5; i++) {
+        if (_dev->read_registers(SPx06_REG_CHIP_ID, &whoami, 1)) {
+        	switch(whoami){
+			case SPL06_CHIP_ID:
+				_spx06_type = BARO_SPL06;
+				is_SPL06=true;
```
Comment (peterbarker) [inline libraries/AP_Baro/AP_Baro_SPx06.cpp:126]:
```suggestion
				type = Type::SPL06;
```

etc

File: libraries/AP_Baro/AP_Baro_SPx06.cpp [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+			case SPA06_CHIP_ID:
+				_spx06_type = BARO_SPA06;
+				is_SPA06=true;
+				break;
+			default:
+				break;
+			}
+        }
+
+		if(is_SPL06 || is_SPA06)
```
Comment (peterbarker) [inline libraries/AP_Baro/AP_Baro_SPx06.cpp:137]:
```suggestion
		if (type != Type::UNKNOWN) {
```

File: libraries/AP_Baro/AP_Baro_SPx06.cpp [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+			break;
+    }
+    
+    if(!is_SPL06 && !is_SPA06) {
```
Comment (peterbarker) [inline libraries/AP_Baro/AP_Baro_SPx06.cpp:141]:
```suggestion
    if (type == Type::UNKNOWN) {
```

File: libraries/AP_Baro/AP_Baro_SPx06.cpp [hunk 7]
```diff
@@ -0,0 +1,316 @@
...
+        return false;
+    }
+
+    // read the calibration data
+    uint8_t SPx06_CALIB_COEFFS_LEN;
+	if(_spx06_type == BARO_SPL06)
```
Comment (peterbarker) [inline libraries/AP_Baro/AP_Baro_SPx06.cpp:147]:
```suggestion
	if(_spx06_type == BARO_SPL06) {
```

... we require braces even on one-line bodies - fix them all, please.

... also an indenting problem

Comment (peterbarker) [inline libraries/AP_Baro/AP_Baro_SPx06.cpp:147]: Use a `switch`

**Review by peterbarker — CHANGES_REQUESTED** [review]
Partial review only.

This is looking good.   I might look at pulling out the renaming into a separate PR as, based on your work here, re-using the same code for the two devices it viable.

**Comment by Hwurzburg** [comment]
does this need an extract_features.py update or will a REGEX catch the change?

File: libraries/AP_Baro/AP_Baro_SPx06.cpp (hunk 7 above)
Comment (radiolinkW) [reply libraries/AP_Baro/AP_Baro_SPx06.cpp]: > Use a `switch`

fixed them all.

**Comment by radiolinkW** [comment]
> does this need an extract_features.py update or will a REGEX catch the change?

It should be updated, I replaced everything related to SPL06 with SPx06.

**Comment by radiolinkW** [comment]
> Partial review only.
> 
> This is looking good. I might look at pulling out the renaming into a separate PR as, based on your work here, re-using the same code for the two devices it viable.

Updated and committed.

**Comment by peterbarker** [comment]
@radiolinkW I've extracted those rename commits over here: https://github.com/ArduPilot/ardupilot/pull/27992/commits - I should be able to get that in soon which will let me take this to DevCallEU tomorrow for a merge, hopefully.

**Comment by radiolinkW** [comment]
> @radiolinkW I've extracted those rename commits over here: https://github.com/ArduPilot/ardupilot/pull/27992/commits - I should be able to get that in soon which will let me take this to DevCallEU tomorrow for a merge, hopefully.

Thanks, won't the other three commits be merged?

**Comment by peterbarker** [comment]
> > @radiolinkW I've extracted those rename commits over here: https://github.com/ArduPilot/ardupilot/pull/27992/commits - I should be able to get that in soon which will let me take this to DevCallEU tomorrow for a merge, hopefully.
> 
> Thanks, won't the other three commits be merged?

Oh yes - but we'll get the other ones in first.  That will make this one much easier to review on a DevCall to get it merged.

**Comment by peterbarker** [comment]
> I also think it should be made into a driver, but the class name should not be SPL06. What do you think the class name should be?SPX06？So the SPL06 driver needs to be rewritten.

Unfortunately we both got this wrong, as it turns out.

tridge was very resistant to renaming this driver, and gave several good reasons why it would be a bad idea - mostly breaking things out-of-tree.  He pointed out several places in our codebase where we simply add comments indicating what devices a driver handles.

I've taken the liberty of removing the rename of the files and the rename of the class commits from your branch and fixing the conflicts

I've force-pushed over your branch here.

If you're happy to proceed from this point I'll leave a review so we can get this merged... also, testing that I haven't broken anything when I resolved those conflicts would be a good idea!

**Comment by radiolinkW** [comment]
> > I also think it should be made into a driver, but the class name should not be SPL06. What do you think the class name should be?SPX06？So the SPL06 driver needs to be rewritten.
> 
> Unfortunately we both got this wrong, as it turns out.
> 
> tridge was very resistant to renaming this driver, and gave several good reasons why it would be a bad idea - mostly breaking things out-of-tree. He pointed out several places in our codebase where we simply add comments indicating what devices a driver handles.
> 
> I've taken the liberty of removing the rename of the files and the rename of the class commits from your branch and fixing the conflicts
> 
> I've force-pushed over your branch here.
> 
> If you're happy to proceed from this point I'll leave a review so we can get this merged... also, testing that I haven't broken anything when I resolved those conflicts would be a good idea!



> > I also think it should be made into a driver, but the class name should not be SPL06. What do you think the class name should be?SPX06？So the SPL06 driver needs to be rewritten.
> 
> Unfortunately we both got this wrong, as it turns out.
> 
> tridge was very resistant to renaming this driver, and gave several good reasons why it would be a bad idea - mostly breaking things out-of-tree. He pointed out several places in our codebase where we simply add comments indicating what devices a driver handles.
> 
> I've taken the liberty of removing the rename of the files and the rename of the class commits from your branch and fixing the conflicts
> 
> I've force-pushed over your branch here.
> 
> If you're happy to proceed from this point I'll leave a review so we can get this merged... also, testing that I haven't broken anything when I resolved those conflicts would be a good idea!

That's also good, there won't be much change.

But I didn't see  a comment in the top of the SPL06 driver saying it supports SPA06 as well, also, RadiolinkPIX6 does not enable Background mode by default.

**Comment by radiolinkW** [comment]
> > I also think it should be made into a driver, but the class name should not be SPL06. What do you think the class name should be?SPX06？So the SPL06 driver needs to be rewritten.
> 
> Unfortunately we both got this wrong, as it turns out.
> 
> tridge was very resistant to renaming this driver, and gave several good reasons why it would be a bad idea - mostly breaking things out-of-tree. He pointed out several places in our codebase where we simply add comments indicating what devices a driver handles.
> 
> I've taken the liberty of removing the rename of the files and the rename of the class commits from your branch and fixing the conflicts
> 
> I've force-pushed over your branch here.
> 
> If you're happy to proceed from this point I'll leave a review so we can get this merged... also, testing that I haven't broken anything when I resolved those conflicts would be a good idea!

That's also good, there won't be much change.

But I didn't see a comment in the top of the SPL06 driver saying it supports SPA06 as well, also, RadiolinkPIX6 does not enable Background mode by default.

File: libraries/AP_Baro/AP_Baro_SPL06.cpp [hunk 8]
```diff
@@ -108,21 +115,44 @@ bool AP_Baro_SPL06::_init()
...
+	switch(type) {
+	case Type::SPL06:
+		SPL06_CALIB_COEFFS_LEN = SPL06_REG_CALIB_COEFFS_END - SPL06_REG_CALIB_COEFFS_START + 1;
+		break;
+	case Type::SPA06:
+		SPL06_CALIB_COEFFS_LEN = SPA06_REG_CALIB_COEFFS_END - SPL06_REG_CALIB_COEFFS_START + 1;
+		break;
+	default:
+		SPL06_CALIB_COEFFS_LEN = 18;
+		break;
```
Comment (peterbarker) [inline libraries/AP_Baro/AP_Baro_SPL06.cpp:153]:
```suggestion
```

by omitting the default case we help someone adding the next `Type::SPx06` driver as they're forced to fill in each of the switch statements.

File: libraries/AP_Baro/AP_Baro_SPL06.cpp [hunk 9]
```diff
@@ -135,12 +165,25 @@ bool AP_Baro_SPL06::_init()
...
     _c30 = ((uint16_t)buf[16] << 8) | (uint16_t)buf[17];
+    if(type == Type::SPA06) {
+		_c31 = (buf[18] & 0x80 ? 0xF000 : 0) | ((uint16_t)buf[18] << 4) | (((uint16_t)buf[19] & 0xF0) >> 4);
+		_c40 = ((buf[19] & 0x8 ? 0xF000 : 0) | ((uint16_t)buf[19] & 0x0F) << 8) | (uint16_t)buf[20];
+	}
 
     // setup temperature and pressure measurements
     _dev->setup_checked_registers(3, 20);
 
+#ifdef SPL06_BACKGROUND_ENABLE
```
Comment (peterbarker) [inline libraries/AP_Baro/AP_Baro_SPL06.cpp:176]:
```suggestion
#if AP_BARO_SPL06_BACKGROUND_ENABLE
```

And ensure this is always define

File: libraries/AP_Baro/AP_Baro_SPL06.cpp [hunk 10]
```diff
@@ -236,8 +292,23 @@ void AP_Baro_SPL06::_update_temperature(int32_t temp_raw)
...
     const float press_raw_sc = (float)press_raw / raw_value_scale_factor(SPL06_PRESSURE_OVERSAMPLING);
-    const float pressure_cal = (float)_c00 + press_raw_sc * ((float)_c10 + press_raw_sc * ((float)_c20 + press_raw_sc * _c30));
-    const float press_temp_comp = _temp_raw * ((float)_c01 + press_raw_sc * ((float)_c11 + press_raw_sc * _c21));
+    float pressure_cal;
+    float press_temp_comp;
+
+    switch(type) {
+	case Type::SPL06:
+		pressure_cal = (float)_c00 + press_raw_sc * ((float)_c10 + press_raw_sc * ((float)_c20 + press_raw_sc * _c30));
+		press_temp_comp= _temp_raw * ((float)_c01 + press_raw_sc * ((float)_c11 + press_raw_sc * _c21));
```
Comment (peterbarker) [inline libraries/AP_Baro/AP_Baro_SPL06.cpp:301]:
```suggestion
		press_temp_comp = _temp_raw * ((float)_c01 + press_raw_sc * ((float)_c11 + press_raw_sc * _c21));
```

File: libraries/AP_Baro/AP_Baro_SPL06.cpp [hunk 11]
```diff
@@ -236,8 +292,23 @@ void AP_Baro_SPL06::_update_temperature(int32_t temp_raw)
...
+		break;
+	case Type::SPA06:
+		pressure_cal = (float)_c00 + press_raw_sc * ((float)_c10 + press_raw_sc * ((float)_c20 + press_raw_sc * ((float)_c30 + press_raw_sc * _c40)));
+		press_temp_comp= _temp_raw * ((float)_c01 + press_raw_sc * ((float)_c11 + press_raw_sc * ((float)_c21) + press_raw_sc * _c31));
```
Comment (peterbarker) [inline libraries/AP_Baro/AP_Baro_SPL06.cpp:305]:
```suggestion
		press_temp_comp = _temp_raw * ((float)_c01 + press_raw_sc * ((float)_c11 + press_raw_sc * ((float)_c21) + press_raw_sc * _c31));
```

File: libraries/AP_Baro/AP_Baro_SPL06.cpp [hunk 12]
```diff
@@ -236,8 +292,23 @@ void AP_Baro_SPL06::_update_temperature(int32_t temp_raw)
 void AP_Baro_SPL06::_update_pressure(int32_t press_raw)
 {
     const float press_raw_sc = (float)press_raw / raw_value_scale_factor(SPL06_PRESSURE_OVERSAMPLING);
-    const float pressure_cal = (float)_c00 + press_raw_sc * ((float)_c10 + press_raw_sc * ((float)_c20 + press_raw_sc * _c30));
-    const float press_temp_comp = _temp_raw * ((float)_c01 + press_raw_sc * ((float)_c11 + press_raw_sc * _c21));
+    float pressure_cal;
+    float press_temp_comp;
```
Comment (peterbarker) [inline libraries/AP_Baro/AP_Baro_SPL06.cpp:296]:
```suggestion
    float pressure_cal = 0;
    float press_temp_comp = 0;
```

File: libraries/AP_Baro/AP_Baro_SPL06.cpp [hunk 13]
```diff
@@ -236,8 +292,23 @@ void AP_Baro_SPL06::_update_temperature(int32_t temp_raw)
...
+		press_temp_comp= _temp_raw * ((float)_c01 + press_raw_sc * ((float)_c11 + press_raw_sc * _c21));
+		break;
+	case Type::SPA06:
+		pressure_cal = (float)_c00 + press_raw_sc * ((float)_c10 + press_raw_sc * ((float)_c20 + press_raw_sc * ((float)_c30 + press_raw_sc * _c40)));
+		press_temp_comp= _temp_raw * ((float)_c01 + press_raw_sc * ((float)_c11 + press_raw_sc * ((float)_c21) + press_raw_sc * _c31));
+		break;
+	default:
+		pressure_cal = 0;
+		press_temp_comp= 0;
+		break;
```
Comment (peterbarker) [inline libraries/AP_Baro/AP_Baro_SPL06.cpp:310]:
```suggestion
```

File: libraries/AP_Baro/AP_Baro_SPL06.cpp [hunk 14]
```diff
@@ -108,21 +115,44 @@ bool AP_Baro_SPL06::_init()
...
+			break;
     }
     
-    if(!is_SPL06) {
+    if (type == Type::UNKNOWN) {
         return false;
     }
 
     // read the calibration data
+    uint8_t SPL06_CALIB_COEFFS_LEN;
```
Comment (peterbarker) [inline libraries/AP_Baro/AP_Baro_SPL06.cpp:143]:
```suggestion
    uint8_t SPL06_CALIB_COEFFS_LEN = 18;
```

**Review by peterbarker — CHANGES_REQUESTED** [review]
Have you tested an SPL06 device to ensure it continues to function normally after these changes?

Please outline the advantages of using the "background" mode of the SPA06...

**Comment by radiolinkW** [comment]
> Have you tested an SPL06 device to ensure it continues to function normally after these changes?
> 
> Please outline the advantages of using the "background" mode of the SPA06...

The latest commit has been tested for narmal flight on the latest Copter4.5 branch.

Background mode only measures continuously, without manually enabling measurements periodically.

**Comment by radiolinkW** [comment]
> @radiolinkW have you tested a device with an SPL06 in it?
> 
> I'm marking this for discussion DevCallEU, and that question _will_ be asked :-)

Tested on RadiolinkPIX6 with SPA06.

**Review by tridge — APPROVED** [review]

**Review by peterbarker — APPROVED** [review]

**Comment by peterbarker** [comment]
Merged, thanks!

**Comment by menschel** [comment]
Hello, does this change also influence the SPL001 update rate?

I have Baro-0 /Baro-4 i.e. Baro Bad Health issues on a plane while circling over a field for 30minutes and I believe it is due to the baro having the same value for 2 seconds.

Meaning to ask, does it make sense to test with AP4.6Beta1 or write an Issue Report directly?

**Pull request merged**